### PR TITLE
make locals.scm queries more granular

### PR DESCRIPTION
**Pull request by the-mikedavis**
I ran into some cases where `locals.scm` would override the highlight for things like record keys and it made me think that

```
(identifier) @local.reference
```

covered too many cases. So this PR makes the reference queries more granular. In general, now the queries try to cover cases where `(identifier)` is being used as an existing binding rather than a pattern or any other syntax element.

I think this covers all the cases now but I'm not too sure :thinking:

### CODE REVIEW

File: test/highlight/destructuring.gleam [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    //          ^ type
+    //              ^ variable
+    //                       ^ variable
+    //                          ^ variable
+  }
+
+  // this is bound above but no longer in scope,
+  // so it should be interpereted as a module
+  z.foo()
+  // <- module
```
Comment (the-mikedavis) [inline test/highlight/destructuring.gleam:23]: this got fixed by the

```
(case_clause) @local.scope
```

addition

File: test/highlight/records.gleam [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  User(name: String)
+  // <- type
+  //    ^ property
+  //          ^ type
+}
+
+pub fn new(name: String) {
+  //        ^ variable.parameter
+  User(name: name)
+  //    ^ property
```
Comment (the-mikedavis) [inline test/highlight/records.gleam:11]: this highlight was mistakenly being overridden to `variable.parameter` which looked quite confusing in the highlights

File: queries/locals.scm [hunk 3]
```diff
@@ -1,6 +1,8 @@
 ; Scopes
 (function_body) @local.scope
 
+(case_clause) @local.scope
```
Comment (J3RN) [inline queries/locals.scm:4]: :100:

This definitely crossed my mind, and then I forgot about it :laughing:

File: queries/locals.scm [hunk 4]
```diff
@@ -10,4 +12,16 @@
 (record_pattern_argument pattern: (identifier) @local.definition)
 
 ; References
-(identifier) @local.reference
+(field_access
+  record: (identifier) @local.reference
+  field: (identifier) @local.reference)
```
Comment (J3RN) [inline queries/locals.scm:17]: Here I don't think `field` can be a local reference as it's the `bar` part of `foo.bar`.

File: test/highlight/destructuring.gleam (hunk 1 above)
Comment (J3RN) [reply test/highlight/destructuring.gleam:23]: :100:

File: queries/locals.scm [hunk 5]
```diff
@@ -10,4 +12,16 @@
...
+
+(function_call
+  function: (identifier) @local.reference)
+
+(argument
+  value: (identifier) @local.reference)
+
+(binary_expression (identifier) @local.reference)
+(case_subjects (identifier) @local.reference)
+(tuple (identifier) @local.reference)
```
Comment (J3RN) [inline queries/locals.scm:27]: This is missing list elements, right? I'm a bit afraid that specifying each of these is a bit error-prone and may quickly become out-of-date as the language evolves. Are we able to specify where an `identifier` is _not_ going to be a reference (e.g. labels) instead?

I'm also open to changing the AST such that labels are actually a different node type, such as `(label)`. This adds a bit of semantic richness to the AST and it should help more than hurt syntax highlighting efforts (I think :sweat_smile:)

**Review by J3RN — COMMENTED** [review]
Good find! :detective: Left a note, curious to hear your thoughts!

File: queries/locals.scm (hunk 5 above)
Comment (the-mikedavis) [reply queries/locals.scm:27]: Aha I knew I forgot one! :sweat_smile: 

I did try something like

```lisp
(field_access field: (identifier) @ignore)
(identifier) @local.reference
```

at first so we would only have to write a list of `(identifier)`s that we _don't_ want locals tracking on, but that doesn't appear to work.

On the one hand I do like a very generic `(identifier)` node in the tree, but I think you might be right: changing the record keys to a different node type would probably make things cleaner. I'll experiment! :scientist:

File: corpus/constants.txt [hunk 6]
```diff
@@ -1,6 +1,6 @@
-==========
+================================================================================
```
Comment (the-mikedavis) [inline corpus/constants.txt:1]: I ran `tree-sitter test --update`, so the diff is a bit ugly from the formatting changes that applies

File: grammar.js [hunk 7]
```diff
@@ -742,6 +742,7 @@ module.exports = grammar({
 
     /* Shared AST nodes */
     identifier: ($) => $._name,
+    label: ($) => $._name,
```
Comment (the-mikedavis) [inline grammar.js:745]: I added this new node `(label)` that takes `(identifier)`'s place in some cases. It really simplifies the `locals.scm` queries :confetti_ball:

File: corpus/constants.txt (hunk 6 above)
Comment (J3RN) [reply corpus/constants.txt:1]: You know, I had wondered where these were coming from 😆

File: queries/highlights.scm [hunk 8]
```diff
@@ -40,19 +48,19 @@
 ; before ":" or after "."
 ; e.g. record field names, tuple indices, names for named arguments, etc
 (field_access
-  field: (identifier) @property)
+  field: (label) @property)
```
Comment (J3RN) [inline queries/highlights.scm:51]: Corresponding to the above description, shouldn't we be able to replace all of the special-case "`(label)` inside `(thing)`" with just one `(label) @property`?

**Review by J3RN — COMMENTED** [review]
I think we may have missed `label` for `_labeled_discard_param` (line 218) and `_labeled_name_param` (221) :thinking:

File: queries/locals.scm (hunk 5 above)
Comment (J3RN) [reply queries/locals.scm]: > I do like a very generic `(identifier)` node in the tree

I must say that I do as well. After reviewing these changes again, I wondered if somehow we were taking a wrong turn. _That said_, I don't think we are. My thoughts on this are probably best embodied by an unrelated comment that I wrote a bit ago:
https://github.com/J3RN/tree-sitter-gleam/blob/b0a23fe88f5f6fa03d87a643e08a54b1aaadbbbd/grammar.js#L474-L475

The name "identifier" even makes perfect sense for "a variable or local function" as it _identifies_ them; perhaps the only more suitable name would be "name", since folks tend to say "function name" or "variable name" rather than "function identifier" or "variable identifier" (probably because those are a mouthful).

The only spots where we deviate from "an identifier is a variable or a function" (with constants, parameters, and assignments targets grouped in with "variables") (disregarding the record-access/remote-function ambiguity) is a couple of spots where we refer to module names as "identifier". However, I would argue that we should probably move to make those use the new `(module)` node type (probably by way of `alias($._name, $.module)` since the actual `$.module` definition is special-purpose).

**TL;DR**: I think we are moving from more ambiguity to less, and I think this is a great asset :100:

Comment (J3RN) [reply queries/locals.scm]: If Gleam changes to allow statements such as this:
```gleam
import gleam/list

const mymod = list
```
Then I'll probably need to rearrange some things such that `(identifier)` means "function name _or_ variable name _or_ module name". We'll see :grin:
